File: python/src/linodemcp/objectdata.py
"""Move object bytes to the presigned URLs the Linode API mints.

This module sits outside ``linodemcp.linode`` deliberately. ``scripts/_routescan.py``
seeds the route scanner on a call named ``request``, so every function reaching
``httpx.AsyncClient.request`` is walked as a route. A PUT to a URL the API handed
back is not a route this client builds, so the transfer below calls ``.put()``
and must never call ``.request(...)``: doing so would report a route the
contract cannot match.

Go's twin is ``go/internal/objectdata``, off its own scanner's surface for the
same reason. The two keep identical defaults and identical refusal sentences,
because the behavior fixtures judge both against one case.
"""

from __future__ import annotations

import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING

import httpx

if TYPE_CHECKING:
    from collections.abc import AsyncIterator

# What an object is stored under when the caller names none. It is filled into
# the presign body rather than onto the PUT, because the signature covers the
# header and a PUT carrying a type the presign request never named is refused.
DEFAULT_CONTENT_TYPE = "application/octet-stream"

# Read size for the streaming upload. Large enough that a multi-gigabyte file
# does not cost a syscall per kilobyte, small enough that peak memory stays flat.
_CHUNK_BYTES = 1024 * 1024


class ObjectDataError(Exception):
    """Base for every transfer failure, so a caller can catch the family."""


class SourceMissingError(ObjectDataError):
    """The source path names no readable file."""


class SourceNotRegularError(ObjectDataError):
    """The source is a directory, device, or socket rather than a file."""


class SourceOutsideRootError(ObjectDataError):
    """The source resolves outside the configured filesystem root."""


class AboveSinglePartCeilingError(ObjectDataError):
    """The file is larger than one presigned PUT carries."""


class TransferError(ObjectDataError):
    """The PUT against the presigned URL failed."""


@dataclass(frozen=True)
class LocalFile:
    """One resolved upload source: the path after symlink resolution, and size."""

    path: Path
    size_bytes: int


@dataclass(frozen=True)
class UploadResult:
    """What the transfer produced, measured off the stream rather than restated.

    ``etag`` arrives verbatim with its quotes stripped and is not checked here.
    For a single PUT the endpoint's ETag is the stored object's MD5, so a caller
    verifies by hashing their own file and comparing, which tests the round trip
    end to end instead of repeating the server's arithmetic back at it.
    """

    etag: str
    size_bytes: int


def inspect(source_path: str, filesystem_root: str) -> LocalFile:
    """Resolve a source path and report what a transfer would send.

    Opens nothing and calls nothing, so the preview and the transfer describe
    the same file. The path is resolved through symlinks before the root
    comparison, so a link inside the root pointing out of it is refused rather
    than followed. An empty root confines nothing, which is the default: a stdio
    server reads the operator's own paths.
    """
    try:
        resolved = Path(source_path).resolve(strict=True)
    except OSError as err:
        msg = f"source_path names no readable file: {source_path}"
        raise SourceMissingError(msg) from err

    _confine(resolved, filesystem_root)

    if not resolved.is_file():
        msg = f"source_path is not a regular file: {source_path}"
        raise SourceNotRegularError(msg)

    return LocalFile(path=resolved, size_bytes=resolved.stat().st_size)


def _confine(resolved: Path, root: str) -> None:
    """Refuse a resolved path outside root. An empty root confines nothing."""
    if not root:
        return

    try:
        resolved_root = Path(root).resolve(strict=True)
    except OSError as err:
        msg = f"source_path resolves outside the configured filesystem root: {resolved}"
        raise SourceOutsideRootError(msg) from err

    # relative_to is the comparison a string prefix gets wrong: /srv/data-other
    # carries /srv/data as a text prefix and is not inside it.
    if not resolved.is_relative_to(resolved_root):
        msg = f"source_path resolves outside the configured filesystem root: {resolved}"
        raise SourceOutsideRootError(msg)


def check_single_part(size_bytes: int, max_bytes: int) -> None:
    """Refuse a file larger than one presigned PUT carries.

    The sentence names the ceiling and the size the caller has, and says what is
    missing rather than naming a tool that does not exist: an object this large
    needs multipart, and Tier A does not implement it.
    """
    if size_bytes <= max_bytes:
        return

    msg = (
        f"source_path is above the single-part upload ceiling: {size_bytes} bytes"
        f" exceeds the {max_bytes} byte ceiling, and multipart upload is not"
        " implemented yet"
    )
    raise AboveSinglePartCeilingError(msg)


async def upload(
    url: str,
    source: LocalFile,
    content_type: str,
    timeout: float,
) -> UploadResult:
    """Stream the source file to the presigned URL and report what landed.

    A fresh client rather than the API client's: the presigned URL carries its
    authorization in the query string, so attaching the Linode token would put
    an account credential on a host that never needed one.

    No error below reports the URL. It is a bearer credential for its lifetime,
    and httpx puts the full URL in its own exception text, which is why the
    failure is reported by type rather than by re-raising what httpx said.
    """
    sent = _ByteCounter()

    try:
        async with httpx.AsyncClient(timeout=timeout) as client:
            response = await client.put(
                url,
                content=_stream(source.path, sent),
                headers={
                    "Content-Type": content_type,
                    # Set explicitly because the body is an iterator with no
                    # length of its own, and a presigned PUT is signed for a
                    # fixed length rather than a chunked one.
                    "Content-Length": str(source.size_bytes),
                },
            )
    except httpx.HTTPError as err:
        msg = "presigned upload failed: the transfer could not be completed"
        raise TransferError(msg) from err

    if not _accepted(response.status_code):
        msg = (
            f"presigned upload failed: the bucket answered HTTP {response.status_code}"
        )
        raise TransferError(msg)

    return UploadResult(
        etag=response.headers.get("etag", "").strip('"'),
        size_bytes=sent.count,
    )


def _refuse_unaccepted(status_code: int) -> None:
    """Refuse a download whose bucket did not answer with a success status."""
    if _accepted(status_code):
        return

    msg = f"presigned download failed: the bucket answered HTTP {status_code}"
    raise TransferError(msg)


def _accepted(status_code: int) -> bool:
    """Report whether the bucket stored the object.

    A presigned PUT answers 200, and some endpoints answer another 2xx, so the
    whole success class counts rather than one exact code.
    """
    return httpx.codes.OK <= status_code < httpx.codes.MULTIPLE_CHOICES


class _ByteCounter:
    """Counts what the stream actually yielded, which is what the result reports."""

    def __init__(self) -> None:
        self.count = 0


async def _stream(path: Path, sent: _ByteCounter) -> AsyncIterator[bytes]:
    """Yield the file in chunks so peak memory stays flat on a large object."""
    with path.open("rb") as handle:
        while chunk := handle.read(_CHUNK_BYTES):
            sent.count += len(chunk)
            yield chunk


def resolved_timeout(seconds: float) -> float:
    """Fall back to the shared 30 minute budget when config left the value at zero.

    Go applies the same fallback in its hook, so a config that names no transfer
    timeout produces one budget rather than two.
    """
    return seconds if seconds > 0 else 1800.0


__all__ = [
    "DEFAULT_CONTENT_TYPE",
    "AboveSinglePartCeilingError",
    "DestinationExistsError",
    "DestinationUnwritableError",
    "DownloadResult",
    "LocalFile",
    "ObjectDataError",
    "SourceMissingError",
    "SourceNotRegularError",
    "SourceOutsideRootError",
    "TransferError",
    "UploadResult",
    "check_single_part",
    "download",
    "inspect",
    "remove",
    "resolve_destination",
    "resolved_timeout",
    "upload",
]


class DestinationExistsError(ObjectDataError):
    """A file is already at the destination and overwrite was not asked for."""


class DestinationUnwritableError(ObjectDataError):
    """The destination cannot be created or written."""


@dataclass(frozen=True)
class DownloadResult:
    """What the download produced, counted off the stream that was written."""

    etag: str
    size_bytes: int


def resolve_destination(
    dest_path: str, filesystem_root: str, *, overwrite: bool
) -> Path:
    """Refuse the two cases the caller has to fix, before any call goes out.

    A file already at the destination without overwrite, and a destination
    outside the configured root. The parent directory is what gets resolved,
    because the file itself does not exist yet.
    """
    target = Path(dest_path)

    if target.exists():
        if not overwrite:
            msg = f"dest_path already exists and overwrite is false: {dest_path}"
            raise DestinationExistsError(msg)
        if target.is_dir():
            msg = f"dest_path cannot be written: {dest_path} is a directory"
            raise DestinationUnwritableError(msg)

    _confine(target.parent.resolve(strict=False), filesystem_root)

    return target


async def download(
    url: str,
    destination: Path,
    timeout: float,
) -> DownloadResult:
    """Stream the presigned URL into destination.

    Written to a temporary file beside the destination and renamed into place
    only after the whole body has landed, so a transfer that fails midway leaves
    no truncated file at the path the caller named.

    Uses ``.stream()`` rather than ``.get()`` so the body never has to be held
    in memory whole; see this module's docstring for why neither may call
    ``.request()``.
    """
    written = 0
    handle = tempfile.NamedTemporaryFile(  # noqa: SIM115
        dir=destination.parent, prefix=".download-", delete=False
    )
    temporary = Path(handle.name)

    try:
        async with (
            httpx.AsyncClient(timeout=timeout) as client,
            client.stream("GET", url) as response,
        ):
            _refuse_unaccepted(response.status_code)
            async for chunk in response.aiter_bytes(_CHUNK_BYTES):
                written += handle.write(chunk)
            etag = response.headers.get("etag", "").strip('"')
    except httpx.HTTPError as err:
        handle.close()
        temporary.unlink(missing_ok=True)
        msg = "presigned download failed: the transfer could not be completed"
        raise TransferError(msg) from err
    except BaseException:
        handle.close()
        temporary.unlink(missing_ok=True)
        raise
    else:
        handle.close()

    temporary.replace(destination)

    return DownloadResult(etag=etag, size_bytes=written)


async def remove(url: str, timeout: float) -> None:
    """Send the DELETE the minted URL authorizes.

    Nothing is measured because nothing moves: the object is either gone
    afterwards or the bucket said why it is not. The request carries no body and
    no Content-Type, the same shape the download's GET sends.

    A fresh client rather than the API client's, for the reason upload gives:
    the presigned URL carries its authorization in the query string. No error
    below reports the URL, since httpx puts the full URL in its own exception
    text and that URL is a bearer credential.
    """
    try:
        async with httpx.AsyncClient(timeout=timeout) as client:
            response = await client.delete(url)
    except httpx.HTTPError as err:
        msg = "presigned removal failed: the request could not be completed"
        raise TransferError(msg) from err

    if not _accepted(response.status_code):
        msg = (
            f"presigned removal failed: the bucket answered HTTP {response.status_code}"
        )
        raise TransferError(msg)

File: go/internal/objectdata/objectdata.go
// Package objectdata moves object bytes to the presigned URLs the Linode API
// mints.
//
// It sits outside internal/linode deliberately. cmd/route-dump resolves the
// requests built in that package against the route contract, and a PUT to a URL
// the API handed back is not a route this client builds: it would surface as
// unresolved route evidence with no contract line able to match it. Keeping the
// transfer here leaves the client package saying only what it declares.
package objectdata

import (
	"context"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// DefaultContentType is what an object is stored under when the caller names
// none. It is filled into the presign body rather than onto the PUT, because
// the signature covers the header and a PUT carrying a type the presign request
// never named is refused.
const DefaultContentType = "application/octet-stream"

// LocalFile is one resolved upload source: the path after symlink resolution
// and the size the transfer expects to send.
type LocalFile struct {
	Path      string
	SizeBytes int64
}

// Inspect resolves sourcePath and reports what a transfer would send, opening
// nothing and calling nothing. The preview and the transfer both read it, so a
// dry run describes the same file the live call would send.
//
// The path is resolved through symlinks before the root comparison, so a link
// inside the root pointing out of it is refused rather than followed. An empty
// filesystemRoot confines nothing: the stdio deployment this feature exists for
// reads the operator's own paths, and a default root would refuse them.
func Inspect(sourcePath, filesystemRoot string) (LocalFile, error) {
	resolved, err := filepath.EvalSymlinks(sourcePath)
	if err != nil {
		return LocalFile{}, fmt.Errorf("%w: %s", ErrSourceMissing, sourcePath)
	}

	if confineErr := confine(resolved, filesystemRoot); confineErr != nil {
		return LocalFile{}, confineErr
	}

	// A stat failure and a non-regular mode answer together: EvalSymlinks above
	// already proved the path resolves, so anything left is "this is not a file
	// the transfer can send", which is the one thing the caller has to fix.
	info, err := os.Stat(resolved)
	if err != nil || !info.Mode().IsRegular() {
		return LocalFile{}, fmt.Errorf("%w: %s", ErrSourceNotRegular, sourcePath)
	}

	return LocalFile{Path: resolved, SizeBytes: info.Size()}, nil
}

// confine refuses a resolved path outside root. An empty root confines nothing.
func confine(resolved, root string) error {
	if root == "" {
		return nil
	}

	resolvedRoot, err := filepath.EvalSymlinks(root)
	if err != nil {
		return fmt.Errorf("%w: %s", ErrSourceOutsideRoot, resolved)
	}

	// filepath.Rel is the comparison a string prefix gets wrong: /srv/data-other
	// carries /srv/data as a text prefix and is not inside it. Anything above the
	// root answers with a leading parent element.
	relative, err := filepath.Rel(resolvedRoot, resolved)
	if err != nil || relative == ".." || strings.HasPrefix(relative, ".."+string(filepath.Separator)) {
		return fmt.Errorf("%w: %s", ErrSourceOutsideRoot, resolved)
	}

	return nil
}

// CheckSinglePart refuses a file larger than one presigned PUT carries. The
// sentence names the ceiling and the size the caller has, and says what is
// missing rather than naming a tool that does not exist: an object this large
// needs multipart, and Tier A does not implement it.
func CheckSinglePart(sizeBytes, maxBytes int64) error {
	if sizeBytes <= maxBytes {
		return nil
	}

	return fmt.Errorf("%w: %d bytes exceeds the %d byte ceiling, and multipart upload is not implemented yet",
		ErrAboveSinglePartCeiling, sizeBytes, maxBytes)
}

// UploadRequest is one transfer against an already-minted presigned URL.
type UploadRequest struct {
	// URL is what the API answered with. It is a bearer credential for its
	// lifetime, which is why no error below ever reports it.
	URL string
	// ContentType must be the value the presign request carried, because the
	// signature covers it.
	ContentType string
	Source      LocalFile
	Timeout     time.Duration
}

// UploadResult is what the transfer produced. SizeBytes is measured off the
// stream rather than restated from the request, so a file that changed size
// mid-read reports what actually reached the bucket.
type UploadResult struct {
	ETag      string
	SizeBytes int64
}

// Upload streams the source file to the presigned URL and reports what landed.
//
// The ETag comes back verbatim with its quotes stripped and is not checked here.
// For a single PUT the endpoint's ETag is the stored object's MD5, so the
// caller verifies by hashing their own file and comparing, which tests the round
// trip end to end instead of repeating the server's arithmetic back at it.
func Upload(ctx context.Context, request UploadRequest) (UploadResult, error) {
	// #nosec G304 -- path is the user-selected local file to upload; reading it is the tool's purpose
	file, err := os.Open(request.Source.Path)
	if err != nil {
		return UploadResult{}, fmt.Errorf("%w: %s", ErrSourceMissing, request.Source.Path)
	}

	defer func() { _ = file.Close() }()

	counter := &countingWriter{}

	response, err := send(ctx, request, io.TeeReader(file, counter))
	if err != nil {
		return UploadResult{}, err
	}

	defer func() { _ = response.Body.Close() }()

	if response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusMultipleChoices {
		return UploadResult{}, fmt.Errorf("%w: the bucket answered HTTP %d", ErrTransfer, response.StatusCode)
	}

	return UploadResult{
		SizeBytes: counter.count,
		ETag:      strings.Trim(response.Header.Get("ETag"), `"`),
	}, nil
}

// send performs the PUT with its own client rather than the API client's: the
// presigned URL carries its authorization in the query string, so attaching the
// Linode token would put an account credential on a host that never needed one.
func send(ctx context.Context, request UploadRequest, body io.Reader) (*http.Response, error) {
	httpRequest, err := http.NewRequestWithContext(ctx, http.MethodPut, request.URL, body)
	if err != nil {
		return nil, fmt.Errorf("%w: the presigned URL could not be used", ErrTransfer)
	}

	httpRequest.Header.Set("Content-Type", request.ContentType)
	// Set explicitly because the body is a reader with no length of its own, and
	// a presigned PUT is signed for a fixed length rather than a chunked one.
	httpRequest.ContentLength = request.Source.SizeBytes

	// Timeout bounds the whole exchange, body included, which is the budget a
	// multi-gigabyte transfer needs and the API client's per-request one cannot
	// give. The context still carries the caller's cancellation.
	response, err := (&http.Client{Timeout: request.Timeout}).Do(httpRequest)
	if err != nil {
		return nil, transferError(ErrTransfer, err)
	}

	return response, nil
}

// transferError strips the URL out of a transport failure and reports it under
// the sentinel the caller's own verb owns. net/http wraps every failure in a
// *url.Error that prints the full URL, and that URL is a bearer credential:
// reporting it would put a transfer token in the tool's error text and from
// there into the audit record.
func transferError(kind, err error) error {
	if urlErr, isURLErr := errors.AsType[*url.Error](err); isURLErr {
		err = urlErr.Err
	}

	return fmt.Errorf("%w: %w", kind, err)
}

// countingWriter counts the bytes the transport actually read off the file,
// which is what the result reports. It is the write half of a tee rather than a
// wrapping reader, so counting costs no extra pass and it has no borrowed error
// to hand back: an io.EOF wrapped on its way to the transport would read as a
// short body and fail the PUT.
type countingWriter struct {
	count int64
}

func (c *countingWriter) Write(data []byte) (int, error) {
	c.count += int64(len(data))

	return len(data), nil
}

// DownloadRequest is one transfer from an already-minted presigned URL to a
// local path.
type DownloadRequest struct {
	// URL is what the API answered with, and a bearer credential for its
	// lifetime, so no error below reports it.
	URL string
	// DestPath is where the object lands. It is confined to FilesystemRoot the
	// same way an upload source is, because a write outside the root is the
	// case a root exists to prevent.
	DestPath       string
	FilesystemRoot string
	Overwrite      bool
	Timeout        time.Duration
}

// DownloadResult is what the transfer produced, counted off the stream that was
// written rather than taken from a header the endpoint sent.
type DownloadResult struct {
	ETag      string
	SizeBytes int64
}

// Download streams the presigned URL into DestPath.
//
// The bytes land in a temporary file beside the destination and are renamed
// into place only after the whole body has been written, so a transfer that
// fails midway leaves no truncated file at the path the caller named. The
// existence check runs before the request, so a refused download costs no call.
func Download(ctx context.Context, request DownloadRequest) (DownloadResult, error) {
	destination, err := ResolveDestination(
		request.DestPath, request.FilesystemRoot, request.Overwrite,
	)
	if err != nil {
		return DownloadResult{}, err
	}

	response, err := fetch(ctx, request)
	if err != nil {
		return DownloadResult{}, err
	}

	// Close drains the unread body itself from Go 1.27, bounded, so a transfer
	// that fails midway drops one pooled connection instead of pulling the
	// discarded remainder of a large object back across the network.
	defer func() { _ = response.Body.Close() }()

	if response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusMultipleChoices {
		return DownloadResult{}, fmt.Errorf("%w: the bucket answered HTTP %d", ErrTransfer, response.StatusCode)
	}

	written, err := writeThroughTemp(destination, response.Body)
	if err != nil {
		return DownloadResult{}, err
	}

	return DownloadResult{
		SizeBytes: written,
		ETag:      strings.Trim(response.Header.Get("ETag"), `"`),
	}, nil
}

// ResolveDestination refuses the two cases the caller has to fix: a file
// already there without overwrite, and a path outside the configured root. The
// parent directory is what gets confined, because the file does not exist yet.
//
// Exported because the download hook calls it before asking for a presigned
// URL: a refused download should cost no request. Download calls it again, so
// the guarantee does not depend on the caller having done so.
func ResolveDestination(destPath, filesystemRoot string, overwrite bool) (string, error) {
	info, err := os.Stat(destPath)
	if err == nil && !overwrite {
		return "", fmt.Errorf("%w: %s", ErrDestinationExists, destPath)
	}

	if err == nil && info.IsDir() {
		return "", fmt.Errorf("%w: %s is a directory", ErrDestinationUnwritable, destPath)
	}

	if confineErr := confine(filepath.Dir(destPath), filesystemRoot); confineErr != nil {
		return "", fmt.Errorf("%w: %s", ErrSourceOutsideRoot, destPath)
	}

	return destPath, nil
}

// fetch performs the GET with its own client, for the reason the upload's PUT
// uses one: the presigned URL carries its authorization in the query string, so
// attaching the Linode token would put an account credential on a host that
// never needed one.
func fetch(ctx context.Context, request DownloadRequest) (*http.Response, error) {
	httpRequest, err := http.NewRequestWithContext(ctx, http.MethodGet, request.URL, http.NoBody)
	if err != nil {
		return nil, fmt.Errorf("%w: the presigned URL could not be used", ErrTransfer)
	}

	response, err := (&http.Client{Timeout: request.Timeout}).Do(httpRequest)
	if err != nil {
		return nil, transferError(ErrTransfer, err)
	}

	return response, nil
}

// writeThroughTemp writes the body to a temporary file in the destination's own
// directory and renames it into place, so the rename is atomic on the same
// filesystem and no partial file is ever visible at the destination.
func writeThroughTemp(destination string, body io.Reader) (int64, error) {
	temporary, err := os.CreateTemp(filepath.Dir(destination), ".download-*")
	if err != nil {
		return 0, fmt.Errorf("%w: %s", ErrDestinationUnwritable, destination)
	}

	written, copyErr := io.Copy(temporary, body)
	closeErr := temporary.Close()

	renameErr := error(nil)
	if copyErr == nil && closeErr == nil {
		renameErr = os.Rename(temporary.Name(), destination)
	}

	// One branch because the three failures are one fact for the caller: nothing
	// landed at the destination. The temporary file goes either way, which is
	// what keeps a failed transfer from leaving a partial file behind.
	if copyErr != nil || closeErr != nil || renameErr != nil {
		_ = os.Remove(temporary.Name())

		return 0, fmt.Errorf("%w: nothing landed at %s", ErrTransfer, destination)
	}

	return written, nil
}

// RemoveRequest is one DELETE against an already-minted presigned URL.
type RemoveRequest struct {
	// URL is what the API answered with, and a bearer credential for its
	// lifetime, so no error below reports it.
	URL string
	// Timeout bounds the whole exchange.
	Timeout time.Duration
}

// Remove sends the DELETE the minted URL authorizes.
//
// Nothing is measured because nothing moves: the object is either gone
// afterwards or the bucket said why it is not. The request carries no body and
// no Content-Type, the same shape the download's GET sends.
func Remove(ctx context.Context, request RemoveRequest) error {
	httpRequest, err := http.NewRequestWithContext(ctx, http.MethodDelete, request.URL, http.NoBody)
	if err != nil {
		return fmt.Errorf("%w: the presigned URL could not be used", ErrRemove)
	}

	// Its own client rather than the API client's, for the reason the upload's
	// PUT gives: the presigned URL carries its authorization in the query
	// string, so attaching the Linode token would put an account credential on
	// a host that never needed one.
	response, err := (&http.Client{Timeout: request.Timeout}).Do(httpRequest)
	if err != nil {
		return transferError(ErrRemove, err)
	}

	defer func() { _ = response.Body.Close() }()

	if response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusMultipleChoices {
		return fmt.Errorf("%w: the bucket answered HTTP %d", ErrRemove, response.StatusCode)
	}

	return nil
}
